core/src/test/java/org/apache/iceberg/TestRewriteManifests.java

```diff
@@ -1105,6 +1108,499 @@ public void testRewriteManifestsOnBranchUnsupported() {
             "Cannot commit to branch someBranch: org.apache.iceberg.BaseRewriteManifests does not support branch commits");
   }
 
+  @Test
+  public void testRewriteDataManifestsPreservesDeletes() {
+    Assumptions.assumeThat(formatVersion).isGreaterThan(1);
+
+    Table table = load();
+
+    table.newAppend().appendFile(FILE_A).appendFile(FILE_B).commit();
+
+    Snapshot appendSnapshot = table.currentSnapshot();
+    Assertions.assertThat(appendSnapshot.dataManifests(table.io())).hasSize(1);
+    Assertions.assertThat(appendSnapshot.deleteManifests(table.io())).isEmpty();
+
+    table.newRowDelta().addDeletes(FILE_A_DELETES).addDeletes(FILE_A2_DELETES).commit();
+
+    Snapshot deleteSnapshot = table.currentSnapshot();
+    Assertions.assertThat(deleteSnapshot.dataManifests(table.io())).hasSize(1);
+    Assertions.assertThat(deleteSnapshot.deleteManifests(table.io())).hasSize(1);
+
+    table.rewriteManifests().clusterBy(file -> file.path().toString()).commit();
```

Comment:
nit: Paths are equivalent here too right? Just wondering about the string

Reply:
It is `ContentFile$path` that returns `CharSequence`. The idea of this closure is to split manifest entries based on the content file path.

Reply:
I meant that "path.toString" and "path" would have equivalent effects in the closure

Reply:
`CharSequence` has no defined `equals` and `hashCode`. You are right it would work without `toString` because `BaseFile` uses `String` internally but it is generally not safe.